### PR TITLE
Refactor dashboard scuemata to introduce composition slots, allow unspecified panels to validate, and re-enable devenv validation

### PR DESCRIPTION
**Pull request by sdboyer**
**Problem:** #38577 turned off validation of devenv dashboards as part of CI as the failures that arose (in #38532) were too confusing to deal with, and it blocked forward progress. That's fine - we need them to be intelligible (per #38712), but we also need a short-term fix so that the particular problem in that case - geomap didn't have a `models.cue` - doesn't end up blocking everything else we want to do.

---

EDIT: THIS IS NO LONGER SHORT-TERM HACKS. It is now 🔥 🔥!

tl;dr - the validator accepts panels without `models.cue`, without having to make the tradeoff on openness for panels having a `models.cue` described in the original approach, below. Really, we just get exactly what we wanted...but also, more!

After the failure of the original approach, it seemed like there wasn't really a good hacky solution at hand. This is now much, much better (despite this being a hard-to-review diff).

Before, we were doing a big chunk of mapping from the panel scuemata form seen in `models.cue`, to the form actually expected by dashboard schema, all of it awkward and in Go, and virtually impossible to replicate directly in CUE.

What i realized is that it makes way, _way_ more sense - especially if we want scuemata composition to be generic, and work from CUE (and i do!) - to have a standard way of exposing "arguments" on a scuemata family for the input scuemata it intends to compose, then have the further processing and productions occur within the dashboard scuemata itself. This realization flipped everything around, and led to the API in this PR. Now, instead of all these nasty mapping layers in Go, [this loop](https://github.com/grafana/grafana/pull/38727/files#diff-b31ceb59c6b295516d72429dcba1a127a0349e060b09523e8c5de43ea9c3d67fR96-R98) is essentially the entirety of what it takes to marry the two scuemata together. And the logic looks essentially the same directly in CUE - it's a one-liner!

```cue

import (
	"github.com/grafana/grafana/packages/grafana-schema/src/scuemata/dashboard"
	"github.com/grafana/grafana/public/app/plugins/panel/timeseries"
)

composedDashboardScuemata: dashboard.Family & { compose: Panel: timeseries: timeseries.Family }
```

Best of all, the way that this works internally within the scuemata family, any Panel scuemata injected in this way will be incorporated into _all_ of the dashboard schemas, so long as future schemas include [these lines](https://github.com/grafana/grafana/pull/38727/files#diff-875a6389294eaa60dc4d37d3de94a40e1a76bac2feb05bedb5ef69408c021053R304-R306). (Maybe we can make the injection happen more automatically, but that's now a schema-authoring-correctness problem, not a schema-consumer-how-do-i-use-this problem - aka, just one nice-to-have option for attaining correctness). That means a CUE consumer can write a file like the above with the logic that composes the `timeseries` plugin scuemata into dashboards once, and then just forget about it, secure in the knowledge that all future references to `composedDashboardScuemata` and any schema within it will have `timeseries` schema composed in, no matter how many times they update the package.

In fact...seeing how easy this is, i suspect we can just do this ourselves. We'll still keep the base dashboard scuemata separate from plugins, but we'll offer an import that does this composition for the CUE user. Then it'll just be up to them to deal with third-party schema.

cc @rgeyer 

---

### Original PR description/hacky approach

There's a crappy, short-term solution to that: make the [`options`](https://github.com/grafana/grafana/blob/a0108a1e5b5639699728629d72f75e4fcbf03b9a/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L224) and [`fieldConfig.defaults.custom`](https://github.com/grafana/grafana/blob/a0108a1e5b5639699728629d72f75e4fcbf03b9a/packages/grafana-schema/src/scuemata/dashboard/dashboard.cue#L296) fields in the schema open - that is, `{...}` instead of `{}`. This will allow JSON from unschematized plugins to pass validation (which is immediately necessary for unfinished core plugins - #38044 - and long-term necessary to handle third-party plugins that may be installed on a Grafana instance), at the cost of forcing open validation of those plugin-specified structs.

To simplify with an example, we're allowing panels without `models.cue` to validate, which currently includes `geomap`. So, a panel containing these values (snipped from larger context):

```json
{
    "type": "geomap",
    "options": {
        "layers": []
    }
}
```

will now pass validation, as we'd want it to once `geomap/models.cue` exists. However, so will this:

```json
{
    "type": "geomap",
    "options": {
        "layers": [],
        "unspecifiedkey": "blah blah blah"
    }
}
```

Because changing `options` to be open means that additional, unspecified fields are allowed. And that also spills over onto other panel types, which _do_ have a `models.cue` defined, like `table`:

```json
{
    "type": "table",
    "options": {
        "showHeader": true,
        "unspecifiedkey": "blah blah blah"
    }
}
```

`showHeader` [is specified](https://github.com/grafana/grafana/blob/a0108a1e5b5639699728629d72f75e4fcbf03b9a/public/app/plugins/panel/table/models.cue#L27), but `unspecifiedKey` is not. Before this PR, the above snippet would have failed validation, but this PR will make it will pass.

I'll be posting an issue about finding a more permanent fix for this - it's tricky - but this unblocks us for now, with an acceptable cost to precise validation in the short term.

### CODE REVIEW

**Review by ryantxu — APPROVED** [review]
This is a good change to get things running again, but I think we should still try to apply open options/config only when we do not have a schema

**Comment by sdboyer** [comment]
lol, whoops. After aaaaall that typing, i should've at least run the test first. In retrospect, it's obvious why this didn't work - there isn't an _actual_ branch in the results of the disjunction where `type: != ""` that can be hit, so we just don't allow unspecified `type`s at all.

i think i've mostly worked out the better solution, anyway, though we may still have to resort to some hacks until certain language features are added to CUE. In any case, will figure out the better approach and push the changes once i do.

**Review by marefr — APPROVED** [review]
LGTM

**Comment by sdboyer** [comment]
i appreciate the LGTM, but this one's gonna undergo some significant refactoring before it actually has the desired effect :)

**Comment by marefr** [comment]
Makes sense, feel free to re-request a review when ready if you want

**Comment by sdboyer** [comment]
Updated PR description and title to reflect the new approach.

i can't overstate how i excited i am that this pulled together so elegantly. The CUE consumer and go consumer logic (can) now mirror each other around a stable, simple interface. It gives us all the runway necessary to make scuemata composition a scalable, fully generic technique. The impossibility of ever achieving generic composition - and the attendant complexity cost for both users and implementors - has been one of my largest existential concerns for scuemata since March.

Even better, doing this also gave me a glimmer of how to handle version and object header injection, which have been another significant question from the outset (n.b. @malcolmholmes).

File: packages/grafana-schema/src/scuemata/dashboard/dashboard.cue [hunk 1]
```diff
@@ -306,7 +300,29 @@ Family: scuemata.#Family & {
...
+                    //   panels with any third-party panel plugins would fail to validate,
+                    //   as well as any core plugins lacking a models.cue. The latter case
+                    //   is not normally expected, but this is not the appropriate place
+                    //   to enforce the invariant, anyway.
+                    // - Instance (this file + core + third-party plugin schema): dashboard
+                    //   json containing panels with a third-party plugin that exists but
+                    //   is not currently installed would fail to validate.
+                    if !list.Contains(compose._panelTypes, type) {
+                        options: {...}
+                        fieldConfig: defaults: custom: {...}
```
Comment (ryantxu) [inline packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:322]: :fire: :taco:

Comment (sdboyer) [reply packages/grafana-schema/src/scuemata/dashboard/dashboard.cue:322]: Yup! Here be the magic.

It was the `list.Contains()` check - and the work necessary to make it possible - that eluded me for a while

**Pull request merged**